### DISCUSSION TITLE
Periodic CSV save

### DISCUSSION THREAD

**Discussion by pin89** [Serial-Studio/Serial-Studio#202] [Ideas]
A super useful function for an amazing logger like that is the periodic (selectable) saving of CSV file with a shadow backup copy. That can be made that application robust against power failures.

Regards, Marco.

**Comment by alex-spataru** [comment]
Thought about that from the start 😄 

The CSV file is saved automatically every second. Serial Studio might crash, or the computer might run out of battery -- but the CSV file will be always updated. Only thing is that a shadow copy is not made, the application writes directly on the CSV file.

Implementation details [here](https://github.com/Serial-Studio/Serial-Studio/blob/master/src/CSV/Export.cpp#L50).

**Reply by pin89** [reply]
Appreciate your reply. In debian environment I find that if I kill Serial Studio during logging the CSV file is there but is has 0 byte size, and therefore is unusable.

![immagine](https://github.com/user-attachments/assets/37d9bbae-4422-4ee6-82b5-b99107663626)

Thanks for reply, anyway.
Marco

**Reply by alex-spataru** [reply]
Thanks Marco, I will do some tests on my system and try to fix this bug. I will keep you updated through this thread.

**Reply by alex-spataru** [reply]
Hi Marco,

I found the source of the problem. I was using a [QTextStream](https://doc.qt.io/qt-6/qtextstream.html) to handle the CSV data, but I wasn’t “flushing” the data to the hard disk until closing the app or disconnecting from the device. Think of it like a toilet—if you don’t flush, everything stays in the bowl (or in this case, a temporary buffer in the RAM) until you take action. This caused the data to remain in memory until the application was closed or the serial device was disconnected.

As described in my last comment, the current approach processes CSV data once per second, but I’ve now made sure that after processing, the data is flushed (written) to the hard disk right away to avoid any loss, even if the app crashes or is killed.

Thanks for the bug report!
